src/Cache/BaseDataCache.php

```diff
@@ -55,10 +55,11 @@ public function get_data(string $key, $default = null)
             return $default;
         }
 
-        // ignore data if internal cache expiration time is expired
-        if ($data['__cache_expiration_time'] < time()) {
-            return $default;
-        }
+        // FreshRSS commented out, to allow HTTP 304
+        // // ignore data if internal cache expiration time is expired
+        // if ($data['__cache_expiration_time'] < time()) {
+        //     return $default;
+        // }
```

Comment:
@Art4 This looks suspicious, probably broken in the current version of SimplePie. This seems to disable the possibility of conditional requests (HTTP 304).

Reply:
This might be referring to https://github.com/simplepie/simplepie/pull/846

Reply:
@Art4 I think the logic is wrong. The cache is destroyed before the conditional request having a chance to kick-in. So you will almost never receive HTTP 304 responses.
For instance, I use a 15-minute cache. Within the next 15 minutes, the cache should be used without any request. After the 15 minutes, a conditional request should be emitted (if information was available), to potentially allow for an HTTP 304 Not Modified response.
This code very much seems to break this, by wrongly destroying the cache after the 15 minutes.

Reply:
This is correct, because this mimics the way PSR-16 works, see https://www.php-fig.org/psr/psr-16/#12-definitions

> **Expiration** - The actual time when an item is set to go stale. This is calculated by adding the TTL to the time when an object is stored.
>
> An item with a 300 second TTL stored at 1:30:00 will have an expiration of 1:35:00.
>
> Implementing Libraries MAY expire an item before its requested Expiration Time, but **MUST treat an item as expired once its Expiration Time is reached**. 

 If `set_data($data, $ttl)` is called with $ttl=15 min, than the cache has to return null (or $default), but not a potential expired value.

Reply:
A workaround would be to set the $ttl to 30 min, but send a request after 15 min and check for HTTP 304 response.

Reply:
Oh, it could be possible that you are mixing up `__cache_expiration_time` and `cache_expiration_time` (note the 2 underscorces). `__cache_expiration_time` is used only internally in BaseDataCache to implement the expiration defined by PSR-16. It uses the $ttl parameter from `set_data($data, $ttl)`.

Inside the stored data there is the `cache_expiration_time` stored. It is defined by SimplePie::fetch_data() in https://github.com/FreshRSS/simplepie/blob/freshrss/src/SimplePie.php#L2071

So what you could do in https://github.com/FreshRSS/simplepie/blob/freshrss/src/SimplePie.php#L2067-L2076

1. keep Line 2071: `'cache_expiration_time' => $this->cache_duration + time(),`
2. Bun increase $ttl in line 2074: 
    ```diff
    - if (!$cache->set_data($cacheKey, $this->data, $this->cache_duration)) {
    + if (!$cache->set_data($cacheKey, $this->data, $this->cache_duration + 300)) {
    ```

Reply:
This sounds wrong 😮
I care way more about the HTTP standard than about PSR-16, and this is not how HTTP conditional requests (HTTP 304) are supposed to work. 
While the cache is valid (which should ideally be defined by the HTTP response `Cache-Control: max-age`), no HTTP request should be performed at all. So your suggestion *send a request after 15 min and check for HTTP 304 response* breaks the HTTP standard.
Later, when the cache is not valid anymore (i.e., way passed your `ttl`, at least as understood in previous versions of SimplePie), a conditional request should be performed. The local cache containing the information for performing a conditional request should be kept for a long time (days).
At the very least, it is a big breaking change of the `ttl` behaviour for existing SimplePie clients.

By the way, related conversation: https://github.com/FreshRSS/FreshRSS/issues/6712
https://rachelbythebay.com/w/2024/08/02/fs/

Reply:
You might be right. Instead of increasing the $ttl one should decrease the `cache_expiration_time` value.

1. Change Line 2071: 
    ```diff
    - 'cache_expiration_time' => $this->cache_duration + time(),`
    + 'cache_expiration_time' => $this->cache_duration - 300 + time(),`
    ```
2. But keep $ttl in line 2074:
    ```
    if (!$cache->set_data($cacheKey, $this->data, $this->cache_duration)) {
    ```

Reply:
If that could help, I can suggest a little online talk, @Art4 ?

Reply:
> Instead of increasing the $ttl one should decrease the `cache_expiration_time` value

Might be a bit better - I need to check, but I doubt this is sufficient. The approach of destroying the cache when accessing it is wrong, or at least very inefficient. The cache should be cleared when it has not been used / accessed for a while.

Reply:
> If that could help, I can suggest a little online talk, @Art4 ?

I'm sorry, but I'm not in a position to do that at the moment.

> > Instead of increasing the $ttl one should decrease the `cache_expiration_time` value
> 
> Might be a bit better - I need to check, but I doubt this is sufficient. The approach of destroying the cache when accessing it is wrong, or at least very inefficient. The cache should be cleared when it has not been used / accessed for a while.

The cache is not destroyed, but ignored. This was added because the deprecated `Base` cache doesn't know the concept of a ttl or cache expiration. But this is nothing you have to deal with if you have a PSR-16 cache implementation. The PSR-16 cache is given a ttl and could also contain a `cache_expiration_time` or even more custom values like `hash` and `mtime` to control the cache handling. If you like I can create a PR with a minimal PSR-16 implementation based on the File cache adapter as a starting point for the caching improvements in FreshRSS.

Reply:
> I'm not in a position to do that at the moment

No worries :-)

> 'cache_expiration_time' => $this->cache_duration - 300 + time(),`

I am not sure where the `300` is coming from, but it should not be hardcoded.

> One can even rename it to `check_rss_feed_after`

Yes, it looks to me like a bit of refactoring would be needed there. I still think the current behaviour is buggy.
Ideally, it should evolve to be HTTP-compliant, which is not the case at the moment, by properly supporting `Cache-Control: max-age` and other related headers.

> or even more custom values

Yes, that would be needed. 

> I can create a PR with a minimal PSR-16 implementation based on the File cache adapter as a starting point

That would be much welcome 👍🏻 It is still unclear to me how much can be put there without changing SimplePie core's code.

> When this clearing is executed is part of the cache implementation and could be triggered by a cronjob

Yes, this is what we have at the moment in FreshRSS (prior to those SimplePie changes), but without needing those extra steps:

https://github.com/FreshRSS/FreshRSS/blob/0f2023811b38fdcbc7d8ade95114ca9403636af5/lib/lib_rss.php#L418-L434

Reply:
> I am not sure where the `300` is coming from, but it should not be hardcoded.

The `300` was only an example.

> > I can create a PR with a minimal PSR-16 implementation based on the File cache adapter as a starting point
> 
> That would be much welcome 👍🏻 It is still unclear to me how much can be put there without changing SimplePie core's code.

Ok, I will create a PR in FreshRSS repo.

> > When this clearing is executed is part of the cache implementation and could be triggered by a cronjob
> 
> Yes, this is what we have at the moment in FreshRSS (prior to those SimplePie changes), but without needing those extra steps:
> 
> https://github.com/FreshRSS/FreshRSS/blob/0f2023811b38fdcbc7d8ade95114ca9403636af5/lib/lib_rss.php#L418-L434

It should be possible to move this logic into the Cache implementation.

Reply:
@Alkarex See https://github.com/Art4/FreshRSS/pull/1 for a PSR-16 implementation based on the file cache driver from SimplePie.

Reply:
Thanks for your efforts 👍🏻

Reply:
@Art4 Could you please change the target of your PR to https://github.com/FreshRSS/simplepie/tree/freshrss
I am not sure I will be able to use it as it does not seem to allow offloading much of my needed patches, while adding 500 lines of boilerplate, but I will try to give it another look later.
In particular, I still fail to see so far how to achieve your claim of *Makes most changes in the cache class of the SimplePie fork not necessary*